### PR TITLE
Prometheus: Add hints for native histograms

### LINKED ISSUE #86388: Show function hints when selecting a native histogram
The query editor will provide hints for classic histogram metrics by showing the function `histogram_quantile`. But, when a user selects a native histogram, none of the native histogram functions are shown as hints.

Classic histogram

Image

Native histogram

Image

Native histogram function hints should include
histogram_avg no documentation link yet, just merged: https://github.com/prometheus/prometheus/pull/13467
[histogram_count](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_count-and-histogram_sum)
[histogram_sum](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_count-and-histogram_sum)
[histogram_fraction](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_fraction)
[histogram_stddev](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_stddev-and-histogram_stdvar)
[histogram_stdvar](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_stddev-and-histogram_stdvar)

How to differentiate classic and native?

If we know a metric is of a certain type, we can show a hint. 

Currently there is no metadata available for native histograms so we have little to differentiate them from classic. (See https://github.com/grafana/alloy/issues/547).

We know classic histograms always end in `_bucket`, `_sum`, or `_count` like 
`cortex_request_duration_seconds_bucket`

And a native histogram metric would have everything but the suffix. 
`cortex_request_duration_seconds`

~~One suggestion is that if we know there is a classic histogram available, we can identify it, remove the suffix and create a collection of native histogram metric names. We could then compare any selected metric to that collection to identify it as a native histogram. This is not as nice as simply checking the metadata for the type, but it is one idea.~~

*update: The above idea will not work because classic histograms will be removed and there will be nothing to compare native histograms to.

~~What we are left with is a metric name with no clear frontend way to identify it's type. Another solution is to create a resource-like call that queries a chosen metric and parses the response to identify the metric as a native histogram, then return this information to the frontend to provide the hint. This would not be scalable to apply to all metrics but it could provide us the hint.~~

**Additional update: use metadata and test with vanilla Prometheus see slack for more context https://raintank-corp.slack.com/archives/C02UC0KCSF4/p1713424061722829?thread_ts=1708519355.394509&cid=C02UC0KCSF4

> We discussed this with Beorn yesterday and the best information and API currently is the [metadata](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-metric-metadata) endpoint. If it has the type==histogram and the series doesn't end in _bucket, _count or _sum, you can be reasonably sure that it is a native histogram.

### PR DESCRIPTION
**Pull request by bohandley**
Fixes https://github.com/grafana/grafana/issues/86388

**What is this feature?**

When a user selects a native histogram, a hint or hints should be provided indicating they have selected a native histogram and providing functions for that metric.

This PR provides hints that allow a user to add 6 functions to the query builder from a native histogram hint

```
histogram_avg(rate(${expression}[$__rate_interval]))
histogram_fraction(0,0.2,rate(${expression}[$__rate_interval]))
histogram_count(rate(${expression}[$__rate_interval]))
histogram_sum(rate(${expression}[$__rate_interval]))
histogram_stddev(rate(${expression}[$__rate_interval]))
histogram_stdvar(rate(${expression}[$__rate_interval]))
```

https://github.com/grafana/grafana/assets/25674746/264ab038-132b-4598-881e-5ae869c07512

The code editor can only show one hint at a time so until we refactor that, we show that the metric is a native histogram and provide the hint to use histogram_avg with the rate function.

Screenshot 2024-04-28 at 9 01 34 PM

Fixes https://github.com/grafana/grafana/issues/86388

**Why do we need this feature?**

Native histograms are being released in Mimir in the next month and this is part of the [GA strategy](https://github.com/grafana/observability-metrics-squad/issues/142).

**Who is this feature for?**

For users who have enabled native histograms.

**Special notes for your reviewer:**

TESTING!!!!!!

1. Use this [repository](https://github.com/laupse/prometheus-native-histograms) which creates native histograms BUT MAKE SURE YOU UPDATE THE FOLLOWING
  - use Prometheus version 2.51.0 int he docker compose file or you won't be able to test all the functions

```
services:
  prometheus:
    image: prom/prometheus:v2.51.0
```
  - update the docker compose Grafana image to use the tag v11.1.0-69868 or later you will have a bad time because you'll need the new parser if you switch between query builder and code editor. 
  - If you don't want to update the grafana image just run Grafana locally with main and create a data source from the prometheus-native-histogram prometheus datasource or add this datasource yaml file.

```
# config file version
apiVersion: 1

# list of datasources to insert/update depending
# what's available in the database
datasources:
  # <string, required> name of the datasource. Required
  - name: Prometheus-native-histogram
    # <string, required> datasource type. Required
    type: prometheus
    # <string, required> access mode. proxy or direct (Server or Browser in the UI). Required
    access: proxy
    # <int> org id. will default to orgId 1 if not specified
    orgId: 1
    # <string> custom UID which can be used to reference this datasource in other parts of the configuration, if not specified will be generated automatically
    uid: prometheus
    # <string> url
    url: http://prometheus:9090
```

2. Now you should have a data source that has a native histogram. The native histogram is named `rpc_durations_native_histogram_seconds`.

3. When you select this in the query builder you will get six query hints (6 functions).
4. When you select this metric in the code editor you will get one hint because the code editor only displays one hint. We need to fix that or create an issue to fix this.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by itsmylife** [comment]
A hint: 
Please make sure you disable `promQLScope` feature flag.

**Review by itsmylife — CHANGES_REQUESTED** [review]
Even I add a the recommended histogram function it keeps suggesting me to add the histogram functions. 

image

Also layout is not the same with your image on my machine. I think some stuff has been changed in time.

**Review by NWRichmond — COMMENTED** [review]
I'm also encountering some issues:

https://github.com/grafana/grafana/assets/5732000/c4d570f4-1f24-4092-b429-f157b3519773

**Comment by zhehao-grafana** [comment]
Thanks, Ismail and Nick, for the review! Can I assume Brenden will be the one fixing these issues when he returns?

**Comment by NWRichmond** [comment]
Hi @zhehao-grafana! Yes, I believe @bohandley will be handling the updates to this PR.

**Comment by bohandley** [comment]
@itsmylife and @NWRichmond thank you for the reviews!

I have made the fix for removing the hints when a native histogram hint is selected. See video below. Because we are removing the hints, we don;t need to change the css so I reverted the css changes and it looks good.

I am updating the video in description with this current video


https://github.com/grafana/grafana/assets/25674746/f4dd441b-7674-40d5-aaf1-95676a872680

**Comment by itsmylife** [comment]
@bohandley Changes looks good. It fixes the issues I mentioned. Thank you. But I have new things :) 

1. In the code mode it only suggests adding `histogram_avg()` function. Should we suggest all possibilities or just one?
2. After having the query in the code mode with a histogram function added, let's say `histogram_avg()`, I switched to builder mode. Then I decided to use some other function so I remove the `histogram_avg()` only. At this point my query is `rate(rpc_durations_native_histogram_seconds[$__rate_interval])`. Then it suggests me all the functions. Which is 👍 . Then if I add `histogram_sum` the query becomes pretty weird. Like this: `histogram_sum(rate(rate(rpc_durations_native_histogram_seconds[$__rate_interval])[$__rate_interval])[$__rate_interval]))`. So it adds rate function once more. And it adds it with a wrong `Range` value. Please see the image below.
image

**Comment by bohandley** [comment]
@itsmylife thank you!

> 1.  In the code mode it only suggests adding histogram_avg() function. Should we suggest all possibilities or just one?

This is a nice to have addition to hints, but is not necessary for pushing native histograms forward at the moment. I will create an issue to allow the code mode to handle multiple hints.

> 2. After having the query in the code mode with a histogram function added, let's say histogram_avg(), I switched to builder mode. Then I decided to use some other function so I remove the histogram_avg() only. At this point my query is rate(rpc_durations_native_histogram_seconds[$__rate_interval]). Then it suggests me all the functions. Which is 👍 . Then if I add histogram_sum the query becomes pretty weird. Like this: histogram_sum(rate(rate(rpc_durations_native_histogram_seconds[$__rate_interval])[$__rate_interval])[$__rate_interval])). So it adds rate function once more. And it adds it with a wrong Range value. Please see the image below.

I'll make a fix for this 👍🏻 . The hints behave by looking for a single metric to make a hint suggestion, so all functions in a hint are added assuming there are no other functions present. Other hints handle this by not offering hints when additional functions are present in the query. Good catch!

**Comment by bohandley** [comment]
Issue for multiple hints here!

https://github.com/grafana/grafana/issues/88564

This is a good first issue!

File: packages/grafana-prometheus/src/query_hints.test.ts [hunk 1]
```diff
@@ -190,4 +190,43 @@ describe('getQueryHints()', () => {
...
+    const mock: unknown = { languageProvider: { metricsMetadata: { foo: { type: 'histogram' } } } };
+    const datasource = mock as PrometheusDatasource;
+
+    let hints = getQueryHints('foo', series, datasource);
+    expect(hints!.length).toBe(6);
+    expect(JSON.stringify(hints)).toContain('ADD_HISTOGRAM_AVG');
+    expect(JSON.stringify(hints)).toContain('ADD_HISTOGRAM_COUNT');
+    expect(JSON.stringify(hints)).toContain('ADD_HISTOGRAM_SUM');
+    expect(JSON.stringify(hints)).toContain('ADD_HISTOGRAM_FRACTION');
+    expect(JSON.stringify(hints)).toContain('ADD_HISTOGRAM_AVG');
```
Comment (NWRichmond) [inline packages/grafana-prometheus/src/query_hints.test.ts:213]: nit: we could do a little less work here by `JSON.stringify`ing once:

```suggestion
    const hintsString = JSON.stringify(hints);
    expect(hintsString).toContain('ADD_HISTOGRAM_AVG');
    expect(hintsString).toContain('ADD_HISTOGRAM_COUNT');
    expect(hintsString).toContain('ADD_HISTOGRAM_SUM');
    expect(hintsString).toContain('ADD_HISTOGRAM_FRACTION');
    expect(hintsString).toContain('ADD_HISTOGRAM_AVG');
```

**Review by NWRichmond — APPROVED** [review]
I appreciate the updates and the helpful code comments. Nicely done, @bohandley!

https://github.com/grafana/grafana/assets/5732000/deb514df-c0ba-486f-a35d-89e8202c45ec

**Comment by itsmylife** [comment]
Sorry for being problem maker :) But there is still a problem. Please check the video:

https://github.com/grafana/grafana/assets/820946/9cb5e6c8-316a-44e4-94b3-d589e2bff6d8

- If there is `rate()` function applied no hints is being shown
- if there is no histogram function applied hints are overflowed

**Comment by zhehao-grafana** [comment]
> Sorry for being problem maker :) But there is still a problem. Please check the video:
> 
>  Screen.Recording.2024-06-03.at.22.58.04.mov 
> * If there is `rate()` function applied no hints is being shown
> * if there is no histogram function applied hints are overflowed

@itsmylife Regarding the first issue [If there is `rate()` function applied no hints is being shown], it is also true for the classic histogram. I don't know whether hints were designed in this way [for example, is hints only applied at the first layer not with a layer in between]

Screenshot by Dropbox Capture

**Comment by itsmylife** [comment]
@zhehao-grafana I missed that. Thank you. In that case, I think the PR is just fine. I'll approve it.

**Review by itsmylife — APPROVED** [review]

**Pull request merged**